pageserver/src/page_service.rs

```diff
@@ -258,7 +266,13 @@ impl<'a> Read for CopyInReader<'a> {
                 Ok(Some(message)) => {
                     let copy_data_bytes = match message {
                         FeMessage::CopyData(bytes) => bytes,
-                        FeMessage::CopyDone => return Ok(0),
+                        FeMessage::CopyDone => {
+                            self.done = true;
+                            return Err(io::Error::new(
+                                io::ErrorKind::UnexpectedEof,
+                                "no more data",
+                            ));
+                        }
```

Comment:
Is the EOF really unexpected here? I guess this relies on the 'tar' code to not try to read any more data from the stream, after reaching the tar end-of-archive marker. But if we used this for anything other than reading a 'tar', receiving CopyDone seems expected.

Even with a tar archive, I would expect us to read the COPY data to completion. If the client sends a valid tar-archive, followed by some garbage, we should not silently ignore the garbage.

Reply:
Yes. There's a problem when we reach end of stream but not end of archive. For example, client sends empty file by mistake.

Is there a regular Eof (not unexpected) I can send somehow?

Reply:
> regular Eof (not unexpected)

With [`std::io::Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read) this is encoded as:
1. not touching the passed in `buf`
2. returning `Ok(0)` as in the bytes read

In addition, it might be better to include a check for empty buffer (however unlikely) in the beginning:

```
if buf.is_empty() {
  return Ok(0);
}

if self.done {
  return Err(...);
}
```

A bit unsure if the returning Err after having returned `Ok(0)` for non-empty `buf` is correct, testing...

Reply:
Continuing to return Ok(0) instead of erroring when `self.done` is at least what `std::net::TcpStream` does, which I think should be the correct way here.